### PR TITLE
EN: 23 create lesson entity

### PR DESCRIPTION
**Pull request by A-Romany**

### CODE REVIEW

File: src/main/java/org/agbrothers/englishbot/processing/ProcessingCore.java [hunk 1]
```diff
@@ -68,7 +68,7 @@ private void applyUserSate(String messageText, User user) {
         switch (messageText) {
             case LinkLabel.MAIN_MENU:
                 user.setStateId(State.MAIN_MENU);
-                lessonService.removeLesson(user.getChatId());
+                //lessonService.removeLesson(user.getLesson());
```
Comment (AgienkoSergiy) [inline src/main/java/org/agbrothers/englishbot/processing/ProcessingCore.java]: why is this here?

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java [hunk 2]
```diff
@@ -10,96 +12,77 @@
...
-        lessonRegistry.put(chatId, lesson);
-        lessonService.setLessonRegistry(lessonRegistry);
 
-        Lesson result = lessonService.getLesson(chatId);
+        when(lessonRepository.findLessonByUser(user)).thenReturn(lesson);
+
+        lessonService.getLesson(user);
 
-        assertEquals(lesson, result);
         verify(wordPoolService, never()).getRandomWordPool();
```
Comment (AgienkoSergiy) [inline src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java:30]: in test signature, it is said `ExpectNoLessonCreation` but you check for not calling `getRandomWordPool`
please, add needed checks

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java [hunk 3]
```diff
@@ -10,96 +12,77 @@
...
+        when(lessonRepository.findLessonByUser(user)).thenReturn(null);
 
-        when(wordPoolService.getRandomWordPool()).thenReturn(words);
+        lessonService.getLesson(user);
 
-        Lesson result = lessonService.getLesson(chatId);
-
-        assertNotNull(result);
-        assertEquals(result, lessonRegistry.get(chatId));
+        verify(lessonRepository, times(1)).saveAndFlush(any(Lesson.class));
```
Comment (AgienkoSergiy) [inline src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: remove a redundant second parameter in `verify` method

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java [hunk 4]
```diff
@@ -10,96 +12,77 @@
...
 
-        Lesson result = lessonService.getLesson(chatId);
+        lessonService.getLesson(user);
 
-        assertNotNull(result);
-        assertNotEquals(lesson, result);
-        assertEquals(lessonRegistry.get(chatId), result);
-        assertNotEquals(lessonRegistry.get(chatId), lesson);
-        verify(wordPoolService).getRandomWordPool();
+        verify(lessonRepository,times(1)).delete(any(Lesson.class));
```
Comment (AgienkoSergiy) [inline src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: same here, `times` matcher is redundant for single call

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java [hunk 5]
```diff
@@ -10,96 +12,77 @@
...
 
         when(wordPoolService.getRandomWordPool()).thenReturn(words);
 
-        Lesson lesson = lessonService.createLesson();
+        lessonService.createLesson(user);
 
         verify(wordPoolService).getRandomWordPool();
-        assertEquals(lesson.getAnswersPool(), words);
-        assertEquals(lesson.getWordPool(), words);
+        verify(lessonRepository,times(1)).saveAndFlush(any(Lesson.class));
```
Comment (AgienkoSergiy) [inline src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: same with `times`

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java [hunk 6]
```diff
@@ -10,96 +12,77 @@
...
-        Lesson lesson = new Lesson(Collections.emptyList(), Collections.emptyList());
-        lessonRegistry.put(chatId, lesson);
-
-        lessonRegistry.remove(chatId);
-
-        assertNull(lessonRegistry.get(chatId));
+        List<Word> words = Collections.singletonList(new Word());
+        User user = new User();
+        lessonService.removeLesson(new Lesson(words, words, user));
+        verify(lessonRepository, times(1)).delete(any(Lesson.class));
```
Comment (AgienkoSergiy) [inline src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: and here

**Review by AgienkoSergiy — CHANGES_REQUESTED** [review]

File: src/main/java/org/agbrothers/englishbot/processing/ProcessingCore.java (hunk 1 above)
Comment (A-Romany) [reply src/main/java/org/agbrothers/englishbot/processing/ProcessingCore.java]: This is my inattention... Thanks for pointing this out.

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java (hunk 2 above)
Comment (A-Romany) [reply src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java:30]: Thanks for comment.

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java (hunk 3 above)
Comment (A-Romany) [reply src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: Thanks for pointing this out.

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java (hunk 4 above)
Comment (A-Romany) [reply src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: and here

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java (hunk 5 above)
Comment (A-Romany) [reply src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: and here

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java (hunk 6 above)
Comment (A-Romany) [reply src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: and here thanks

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java [hunk 7]
```diff
@@ -10,96 +12,98 @@
...
+        lessonService.createLesson(user);
 
         verify(wordPoolService).getRandomWordPool();
-        assertEquals(lesson.getAnswersPool(), words);
-        assertEquals(lesson.getWordPool(), words);
+        verify(lessonRepository).saveAndFlush(any(Lesson.class));
+    }
+
+    @Test
+    void removeLesson_NoUser_ExpectRemoveUser() {
```
Comment (AgienkoSergiy) [inline src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: the name of the test is not understandable

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java [hunk 8]
```diff
@@ -10,96 +12,98 @@
...
     }
 
     @Test
-    void removeLesson() {
-        String chatId = "testChatId";
-        Map<String, Lesson> lessonRegistry = new HashMap<>();
-        lessonService.setLessonRegistry(lessonRegistry);
-        Lesson lesson = new Lesson(Collections.emptyList(), Collections.emptyList());
-        lessonRegistry.put(chatId, lesson);
+    void removeLesson_HasUser_ExpectRemoveUser() {
```
Comment (AgienkoSergiy) [inline src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: same thing with signature here

**Review by AgienkoSergiy — CHANGES_REQUESTED** [review]

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java (hunk 8 above)
Comment (A-Romany) [reply src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: Thank you. Checked

File: src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java (hunk 7 above)
Comment (A-Romany) [reply src/test/java/org/agbrothers/englishbot/service/LessonServiceTest.java]: Thanks. Done

**Review by AgienkoSergiy — APPROVED** [review]
